pkg/restore/restore.go

```diff
@@ -259,50 +259,98 @@ func (ctx *context) execute() (api.RestoreResult, api.RestoreResult) {
-	exists, err := ctx.fileSystem.DirExists(clusterPath)
-	if err != nil {
-		errors.Cluster = []string{err.Error()}
-	}
-	if exists {
-		w, e := ctx.restoreNamespace("", clusterPath)
-		merge(&warnings, &w)
-		merge(&errors, &e)
-	}
+	namespaceFilter := collections.NewIncludesExcludes().Includes(ctx.restore.Spec.IncludedNamespaces...).Excludes(ctx.restore.Spec.ExcludedNamespaces...)
 
-	// namespace-scoped
-	namespacesPath := path.Join(dir, api.NamespaceScopedDir)
-	exists, err = ctx.fileSystem.DirExists(namespacesPath)
+	// Make sure the top level "resources" dir exists:
+	resourcesDir := path.Join(dir, api.ResourcesDir)
+	_, err := ctx.fileSystem.DirExists(resourcesDir)
 	if err != nil {
 		addArkError(&errors, err)
 		return warnings, errors
 	}
-	if !exists {
-		return warnings, errors
-	}
 
-	nses, err := ctx.fileSystem.ReadDir(namespacesPath)
+	resourceDirs, err := ctx.fileSystem.ReadDir(resourcesDir)
 	if err != nil {
 		addArkError(&errors, err)
 		return warnings, errors
 	}
 
-	namespaceFilter := collections.NewIncludesExcludes().Includes(ctx.restore.Spec.IncludedNamespaces...).Excludes(ctx.restore.Spec.ExcludedNamespaces...)
-	for _, ns := range nses {
-		if !ns.IsDir() {
+	resourceDirsMap := make(map[string]os.FileInfo)
+
+	for _, rscDir := range resourceDirs {
+		rscName := rscDir.Name()
+		resourceDirsMap[rscName] = rscDir
+	}
+
+	for _, resource := range ctx.prioritizedResources {
+		rscDir := resourceDirsMap[resource.String()]
+		if rscDir == nil {
 			continue
 		}
-		nsPath := path.Join(namespacesPath, ns.Name())
 
-		if !namespaceFilter.ShouldInclude(ns.Name()) {
-			ctx.infof("Skipping namespace %s", ns.Name())
-			continue
+		resourcePath := path.Join(resourcesDir, rscDir.Name())
+
+		clusterSubDir := path.Join(resourcePath, api.ClusterScopedDir)
+		clusterSubDirExists, err := ctx.fileSystem.DirExists(clusterSubDir)
+		if err != nil {
+			addArkError(&errors, err)
+			return warnings, errors
+		}
+		if clusterSubDirExists {
+			w, e := ctx.restoreResource(resource.String(), "", clusterSubDir)
+			merge(&warnings, &w)
+			merge(&errors, &e)
+		}
+
+		nsSubDir := path.Join(resourcePath, api.NamespaceScopedDir)
+		nsSubDirExists, err := ctx.fileSystem.DirExists(nsSubDir)
+		if err != nil {
+			addArkError(&errors, err)
+			return warnings, errors
 		}
+		if nsSubDirExists {
+			nsDirs, err := ctx.fileSystem.ReadDir(nsSubDir)
+			if err != nil {
+				addArkError(&errors, err)
+				return warnings, errors
+			}
+
+			for _, nsDir := range nsDirs {
+				if !nsDir.IsDir() {
+					continue
+				}
+				nsPath := path.Join(nsSubDir, nsDir.Name())
+
+				if !namespaceFilter.ShouldInclude(nsDir.Name()) {
+					ctx.infof("Skipping namespace %s", nsDir.Name())
+					continue
+				}
 
-		w, e := ctx.restoreNamespace(ns.Name(), nsPath)
-		merge(&warnings, &w)
-		merge(&errors, &e)
+				nsName := nsDir.Name()
+
+				// fetch mapped NS name
+				// TODO: what is a mapped ns name? for renames?
```

Comment:
This is when you do `ark restore create --namespace-mappings foo:bar,hello:goodbye,...`. Any resources in `foo` will be restored to `bar` instead, etc.

Reply:
Thanks, I think we might need to make use of this at some point.